src/Polynomials.jl

```diff
@@ -363,6 +365,7 @@ p(4)   # 57 = 1 + 2*4 + 3*4^2
 """
 function polyval{T,S}(p::Poly{T}, x::S)
     R = promote_type(T,S)
+    T<:AbstractFloat && any(isnan(p.a)) && convert(R, NaN)
```

Comment:
Actually, both. It is very ugly, as you have already agreed on :) Moreover, it brings yet another `if` check into action --- after all, function parameter `T` is known apriori. We should avoid any last minute decision making, which means we should not put a check on `T`, ideally. If you check Julia `Base`, `isnan` is implemented as follows:

```julia
isnan(x::Real) = false
isnan(x::Float64) = # some check
# same for `isnan(x::Float32)` and `isnan(x::Float16)`
isnan(x::Complex) = isnan(x.re) | isnan(x.im)
```

Looking at `Base`, we can assume any `Number` type should implement its own `isnan`. As in Keno's example for `Mod2`, either he needs to implement `isnan` for his type, or, he will define the type as `Mod2 <: Real` instead of `Mod2 <: Number` to borrow from `Real`'s implementations. As for your types (i.e., `Poly` coefficients), I have already implemented `isnan`.

In short, to me, we should drop `T<:AbstractFloat` case and assume `isnan` should be defined for all custom `T<:Number` types.

Any opposition?